pkg/actors/internal/placement.go

```diff
@@ -0,0 +1,374 @@
+
+				newStream, newConn := p.establishStreamConn()
+				if newStream != nil {
+					p.clientConn = newConn
+					p.clientStream = newStream
+					p.streamConnectedCh <- true
+					p.streamConnAlive = true
+				}
+
+				continue
+			}
+
+			diag.DefaultMonitoring.ActorStatusReported("recv")
+			p.onPlacementOrder(resp)
+		}
+	}()
+
+	// Send the current host status to placement to register the member and
+	// maintain the status of member by placement.
+	p.shutdownConnLoop.Add(1)
+	go func() {
+		defer p.shutdownConnLoop.Done()
+		for !p.shutdown {
+			// Wait until stream is reconnected.
+			if !p.streamConnAlive || p.clientStream == nil {
+				<-p.streamConnectedCh
+			}
+			if p.shutdown {
+				break
+			}
+
+			// appHealthFn is the health status of actor service application. This allows placement to update
+			// memberlist and hashing table quickly.
+			if !p.appHealthFn() {
+				// app is unresponsive, close the stream and disconnect from the placement service.
+				// Then Placement will remove this host from the member list.
+				log.Debug("disconnecting from placement service by the unhealthy app.")
+
+				p.closeStream()
+				continue
+			}
+
+			host := v1pb.Host{
+				Name:     p.runtimeHostName,
+				Entities: p.actorTypes,
+				Id:       p.appID,
+				Load:     1, // Not used yet
+				// Port is redundant because Name should include port number
+			}
+
+			err := p.clientStream.Send(&host)
+			if err != nil {
+				diag.DefaultMonitoring.ActorStatusReportFailed("send", "status")
+				log.Debugf("failed to report status to placement service : %v", err)
+			}
+
+			time.Sleep(statusReportHeartbeatInterval)
+		}
+	}()
+}
+
+// Stop shuts down server stream gracefully.
+func (p *ActorPlacement) Stop() {
+	p.shutdown = true
+	close(p.streamConnectedCh)
+	p.closeStream()
+	p.shutdownConnLoop.Wait()
+}
+
+func (p *ActorPlacement) closeStream() {
+	if p.clientStream != nil {
+		p.clientStream.CloseSend()
+	}
+
+	if p.clientConn != nil {
+		p.clientConn.Close()
+	}
+}
+
+func (p *ActorPlacement) establishStreamConn() (v1pb.Placement_ReportDaprStatusClient, *grpc.ClientConn) {
+	for !p.shutdown {
+		serverAddr := p.serverAddr[p.serverIndex]
+
+		// Stop reconnecting to placement until app is healthy.
+		if !p.appHealthFn() {
+			time.Sleep(placementReconnectInterval)
+			continue
+		}
+
+		log.Infof("try to connect to placement service: %s", serverAddr)
+
+		opts, err := dapr_credentials.GetClientOptions(p.clientCert, security.TLSServerName)
+		if err != nil {
+			log.Errorf("failed to establish TLS credentials for actor placement service: %s", err)
+			return nil, nil
+		}
+
+		if diag.DefaultGRPCMonitoring.IsEnabled() {
+			opts = append(
+				opts,
+				grpc.WithUnaryInterceptor(diag.DefaultGRPCMonitoring.UnaryClientInterceptor()))
+		}
+
+		conn, err := grpc.Dial(serverAddr, opts...)
+	NEXT_SERVER:
+		if err != nil {
+			log.Debugf("error connecting to placement service: %v", err)
+			conn.Close()
```

Comment:
Can this be `nil` when `err != nil`? May want to guard with nil check to avoid panic.

Reply:
that's good suggestion.  will do